### PR TITLE
:sparkles: Add events callbacks to ECS

### PR DESCRIPTION
**Pull request by C0Florent**
This PR adds various event hooks in our ECS. The four types of events are:
- Entity created
- Entity killed
- Component created
- Component removed

The added methods to register callbacks related to entities are in our `Registry`, whereas the methods to register callbacks related to components are in `ComponentManager`. Feel free to add "bridge" methods in the `Registry` or even in `GameEngine` to call the lower-level methods if you feel like it.

Another improvement brought by this PR is the ability to delete all components from a given entity, which is automatically done when calling `Registry::killEntity`.

### CODE REVIEW

File: src/ECS/registry/Registry.cpp [hunk 1]
```diff
@@ -0,0 +1,21 @@
+#include "Registry.hpp"
```
Comment (Tugduoff) [inline src/ECS/registry/Registry.cpp:1]: Missing header in registry.cpp file ^^

File: src/ECS/registry/Registry.hpp [hunk 2]
```diff
@@ -9,8 +9,11 @@
     #define REGISTRY_HPP
 
     #include "ECS/component/ComponentManager.hpp"
+    #include "ECS/entity/Entity.hpp"
     #include "ECS/entity/EntityManager.hpp"
     #include "ECS/system/SystemManager.hpp"
+#include <functional>
```
Comment (Tugduoff) [inline src/ECS/registry/Registry.hpp:15]: Missing tab

File: src/ECS/utilities/SparseArray.hpp [hunk 3]
```diff
@@ -127,13 +138,47 @@ class SparseArray {
...
+         * @brief each time a component is removed
+         */
+        void registerRemoveCallback(RemoveCallback cb)
+        {
+            __removeCallbacks.push_back(std::move(cb));
         }
 
     private:
 
+        void __callAllCreate(size_t index) {
```
Comment (Tugduoff) [inline src/ECS/utilities/SparseArray.hpp:164]: Could you add a doxygen comment for this function and the following one ?

**Review by Tugduoff — CHANGES_REQUESTED** [review]

**Review by Tholarr — DISMISSED** [review]
Nothing more to say than Tug's comments :saluting_face:

**Review by Tugduoff — APPROVED** [review]

**Review by Tholarr — APPROVED** [review]
:+1:

**Pull request merged**